File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    n = len(numbers)
    answer = 0
    def dfs(level, result):
        if level==n:
            if result==target:
                nonlocal answer
                answer += 1
            return
        else:
            dfs(level+1, result+numbers[level])
            dfs(level+1, result-numbers[level])
    dfs(0, 0)
        
    return answer
    

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    
    static int count = 0;
    public int solution(int[] numbers, int target) {
        dfs(0, numbers, target, new char[numbers.length], 0);
        return count;
    }
    
    public void dfs(int level, int[] numbers, int target, char[] op, int result) {
        if (level==numbers.length) {
            if (result==target) {
                count++;
            }
            return;
        }
        for (int i = 0; i<2; i++) {
            if (i==0) {
                op[level] = '+';
                dfs(level+1, numbers, target, op, result+numbers[level]);
            }
            else if (i==1) {
                op[level] = '-';
                dfs(level+1, numbers, target, op, result-numbers[level]);
            }
        }
    }
}